kernel/init.c

```diff
@@ -138,3 +138,25 @@ int create_stdio(const char *file) {
     current->files->files[2] = fd_retain(fd);
     return 0;
 }
+
+static struct fd *open_fd_from_actual_fd(int fd_no) {
+    if (fd_no < 0)
+        return ERR_PTR(errno_map());
+    struct fd *fd = fd_create(&realfs_fdops);
+    fd->real_fd = fd_no;
+    fd->dir = NULL;
+    return fd;
+}
+
+int create_piped_stdio() {
+    struct fd *si = open_fd_from_actual_fd(STDIN_FILENO);
+    struct fd *so = open_fd_from_actual_fd(STDOUT_FILENO);
+    struct fd *se = open_fd_from_actual_fd(STDERR_FILENO);
+    si->refcount = 0;
+    so->refcount = 0;
+    se->refcount = 0;
+    current->files->files[0] = fd_retain(si);
+    current->files->files[1] = fd_retain(so);
+    current->files->files[2] = fd_retain(se);
+    return 0;
```

Comment:
This would be simpler like `current->files->files[0] = open_fd_from_actual_fd(STDIN_FILENO)` etc. The refcount when the file is returned is just 1.

Also should check for if open_fd_from_actual_fd returns an error and pass it on.

Reply:
How's this

Reply:
It's gonna need to use the IS_ERR family of macros.

Reply:
Update: it isn't, adhoc_fd_create just returns NULL instead of ERR_PTR(_ENOMEM).